### PR TITLE
Add kubernetes registry in front of ServiceEntry

### LINKED ISSUE #50968: Sidecar CRD is ignored when pod rebalance from one Istiod to another and is endpoint in a ServiceEnrty
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

We have a Sidecar CRD with configurations of `egress.hosts` that match pods with sidecar injected by `workloadSelector`.
We also create a ServiceEntry and add those pods as endpoints in a ServiceEntry by IP address.

Once we do it, we see that 30 minutes after the sidecars boot (when the sidecar connection to Istiod is recreated), Istiod ignores the Sidecar CRD and the egress configurations and starts to push all the mesh configs to the sidecars.
```
{"level":"info","time":"2024-05-09T18:54:04.373168Z","scope":"ads","msg":"CDS: PUSH for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:8 size:3.1kB cached:4/4"}
{"level":"info","time":"2024-05-09T18:54:04.373897Z","scope":"ads","msg":"LDS: PUSH for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:6 size:9.1kB"}
{"level":"info","time":"2024-05-09T18:54:04.373947Z","scope":"ads","msg":"RDS: PUSH for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:2 size:1.2kB cached:2/2"}
{"level":"info","time":"2024-05-09T18:54:04.374017Z","scope":"ads","msg":"NDS: PUSH for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:1 size:92B"}
{"level":"info","time":"2024-05-09T19:13:59.430100Z","scope":"ads","msg":"ADS: \"10.120.19.17:36808\" sleep-ilo-2-9ff6468bc-5bl8n.devops-apps-348 terminated"}
{"level":"info","time":"2024-05-09T19:13:59.699261Z","scope":"ads","msg":"ADS: new connection for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps-357"}
{"level":"info","time":"2024-05-09T19:13:59.700163Z","scope":"ads","msg":"CDS: PUSH request for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:154 size:91.9kB cached:139/151"}
{"level":"info","time":"2024-05-09T19:13:59.700397Z","scope":"ads","msg":"EDS: PUSH request for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:4 size:1.3kB empty:0 cached:4/4"}
{"level":"info","time":"2024-05-09T19:13:59.702351Z","scope":"ads","msg":"LDS: PUSH request for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:152 size:169.7kB"}
{"level":"info","time":"2024-05-09T19:13:59.702907Z","scope":"ads","msg":"NDS: PUSH request for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:1 size:19.8kB"}
{"level":"info","time":"2024-05-09T19:13:59.702960Z","scope":"ads","msg":"RDS: PUSH request for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:2 size:1.2kB cached:2/2"}
{"level":"info","time":"2024-05-09T19:13:59.742769Z","scope":"ads","msg":"EDS: PUSH request for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:122 size:53.4kB empty:0 cached:114/122"}
{"level":"info","time":"2024-05-09T19:13:59.866555Z","scope":"ads","msg":"RDS: PUSH request for node:sleep-ilo-2-9ff6468bc-5bl8n.devops-apps resources:68 size:74.1kB cached:48/68"}
```

It happens even if the connection is recreated to the same Istiod.
We tried setting the ServiceEntry.spec.location to MESH_EXTERNAL and MESH_INTERNAL, but it didn’t change.

This can also be reproduced by deleting the Istiod that the sidecar is connected to:
```
 ~  istioctl pc endpoints -n devops-apps sleep-ilo-2-9ff6468bc-m2jl2 | wc -l
      13
 ~  istioctl ps | grep sleep-ilo-2-9ff6468bc-m2jl2
sleep-ilo-2-9ff6468bc-m2jl2.devops-apps         general-dev-02     SYNCED     SYNCED     SYNCED     SYNCED     istiod-77b697dd85-m929j     1.21.1
 ~  kubectl -n istio-system delete pods istiod-77b697dd85-m929j
pod "istiod-77b697dd85-m929j" deleted
 ~  istioctl pc endpoints -n devops-apps sleep-ilo-2-9ff6468bc-m2jl2 | wc -l
     321
 ~ 
```

### Version

```Text
istioctl version
client version: 1.13.5
control plane version: 1.21.1
data plane version: 1.21.1 (15 proxies)

kubectl version
Client Version: v1.29.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.3-eks-adc7111
```

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Fix #50968

Without this we can get proxy labels from WLE when the wle's address = pod ip

### CODE REVIEW

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 1]
```diff
@@ -139,7 +139,24 @@ func NewController(opt Options) *Controller {
 }
 
 func (c *Controller) addRegistry(registry serviceregistry.Instance, stop <-chan struct{}) {
-	c.registries = append(c.registries, &registryEntry{Instance: registry, stop: stop})
+	added := false
+	if registry.Provider() == provider.Kubernetes {
+		for i, r := range c.registries {
+			if r.Provider() != provider.Kubernetes {
+				// insert the registry in the position of the first non kubernetes registry
+				newRegistries := make([]*registryEntry, len(c.registries)+1)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/aggregate/controller.go:147]: nit: can we use slices.Insert to avoid this complex code

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/aggregate/controller.go:147]: good to know that function.

**Review by ramaraochavali — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

**Comment by costinm** [comment]
Any impact on backwards compatibility ?

Should we prevent WE from using the same IP as a pod instead ? Any security
implications ?

**Comment by costinm** [comment]
On Tue, May 14, 2024 at 7:38 PM hzxuzhonghu ***@***.***>
wrote:

> Should we prevent WE from using the same IP as a pod instead ? Any security
> implications ?
>
> That can not be 100% possible.
>
We do build a map from IP to 'workload info' - so it should not be
impossible to detect.
Or giving users a mechanism to set CIDR ranges allowed for WE ( assigned to
their VMs
and separate from the Pod/Services CIDR).

To be fair - Endpoints and EndpointSlices don't seem to have any protection
either. I don't remember
if we still consume them (for the 'custom' EndpointSlices and selector-less
cases ).


> Any impact on backwards compatibility ?
>
> @costinm <https://github.com/costinm> Maybe no, and in services ordering
> we have already sort kubernetes service before serviceentry
>
I see, so this will sort Pods before WE ?

I think it's reasonable - we don't make guarantees and it's probably
something wrong if users set pod IPs in WE.

+1 from me.

**Comment by hzxuzhonghu** [comment]
> We do build a map from IP to 'workload info' - so it should not be
impossible to detect.

OK, i know your point. We can do that in istiod. I meant we cannot prevent it from creating with validating webhook

**Comment by hzxuzhonghu** [comment]
It seems out certificate used for test has expired again

```
{Failed      utils_test.go:312: failed at updateMutatingWebhookConfig: failed to verify the certificate chain: x509: certificate has expired or is not yet valid: current time 2024-05-15T02:46:41Z is after 2024-05-14T23:11:26Z}

```

**Pull request merged**